### DISCUSSION TITLE
osu!mania Hit Bursts Animation Frame Limit

### DISCUSSION THREAD

**Discussion by DLGEMS** [ppy/osu#18109] [Q&A] [answered]
I want to animate the osu!mania's hit burst, i.e., the text that says 300, 200, 100, 50.
The [osu!mania skinning wiki page](https://osu.ppy.sh/wiki/en/Skinning/osu%21mania) mentions that the hit bursts are animatable.

However, I can not make an extended animation lasting more than a couple of frames.
There seems to be a limit of around 4 frames before osu! just refuses to display the next upcoming frames (e.g., `mania-hit300g-5@2x.png`, or `-6`, `-7`)

Is there a way that I can bypass this limit?

**Comment by bdach** [comment]
I'm not aware of any such limitations being in place. At least not in lazer.

**Reply by DLGEMS** [reply]
Thank you for your reply.

That is really strange since it also did successfully plays the first few frames.
The default "popup" animation imposed by osu! where the sizes are varied is still in effect.
I even made sure that the frame images have the exact same dimensions.

By any chance do you know any osu!mania skins where they were able to extend the animations more than these first frames?
Or the ones where that is able to successfully override the default "popup" animation?

**Comment by peppy** [comment, marked as answer]
Please provide your skin showing this not working. Also make sure you aren't missing a frame in the sequence.

**Reply by DLGEMS** [reply]
Thanks for looking into this. I have ensured that there is not a missing frame in the sequence.
Please find the download link to the skin [here](https://github.com/ppy/osu/files/8638450/DJMAX.Genesis.Light.DLGEMS.edit.zip).

I have not gone ahead and fully animate this yet. Right now its only a play on the text size and transparency.
The expected behavior is for the hit bursts (the text MAX 100%, etc.) to pop up (increase in size), linger a bit longer, then fade out.

Since the default pop up animation (as seen in all hit burst animation in [this video](https://youtu.be/eKKfObEUiik?t=21)) is still in effect, it is clashing with the expected behavior.

**Reply by peppy** [reply]
First of all, you shouldn't have the non `@2x` versions since they are the same size as the `@2x` ones. Either delete them or actually export at half size (note that this isn't required for lazer - lazer will only read the `@2x` ones for now).

As for why this isn't working: you're trying to add animations via textures. Animations are hard coded in the game at the moment so you can't do this. If you want the animations to play faster (to have them play before they fade due to the game animations) then adjust `AnimationFramerate` in skin.ini.

You can see the animations being applied by the game here:

https://github.com/ppy/osu/blob/b907c2f4f6bb37c22428520c238b7d6861ec5fde/osu.Game.Rulesets.Mania/UI/DrawableManiaJudgement.cs#L48-L53

This is the same in stable and lazer. We will allow more flexibility over this in the future, but right now all skinning is done in line with legacy (aka osu-stable) standards. You can learn more on the wiki.

This is not a bug, though, so I'll mark this as the answer.

**Reply by DLGEMS** [reply]
Thank you for your explanation. I look forward to this feature being implemented.
I was not careful while generating the animations. I'll ensure that the SD and HD textures are correctly sized.

From my understanding, this would mean that only the textures that can be drawn within 50ms delay will be displayed, right?
Since [it is mentioned](https://osu.ppy.sh/wiki/en/Skinning/osu%21mania#hit-bursts) that the hit bursts use a fixed 60fps animation frame rate (16.67ms frame time), it would make sense that only a few of them will be drawn within that time period.

Another small question, does the game still render additional textures during the 200ms fade out?

**Reply by peppy** [reply]
It does, yes
